ccmlib/common.py

```diff
@@ -785,9 +786,12 @@ def parse_settings(args):
             val = False
         else:
             try:
-                val = int(val)
-            except ValueError:
+                val = ast.literal_eval(str(val))
```

Comment:
what if we have something like: `hello: "world"` here? before this change, the `val` is `'"world"'`, after this change the `val` would be `'world'`. and what if we really want to set a value like `'1e10'`. i still cannot tell if this is a valid concern after reading 
https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L832-L841 and https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L872-L875

or do we accept `foo: [1, False, "hello"]` ?

actually, i was about to suggest using `json.loads()` but i just realized that we are not actually expecting json encoded data here. it's kind of freestyle format. but i still think it would be better to stick to a well-defined format / encoding, so we don't need to worry about these things.

but if we don't really care. i'd suggest at least handle more types of exceptions, see https://docs.python.org/3/library/ast.html#ast.literal_eval

Reply:
I think

> what if we have something like: `hello: "world"` here? before this change, the `val` is `'"world"'`, after this change the `val` would be `'world'`. and what if we really want to set a value like `'1e10'`. i still cannot tell if this is a valid concern after reading

regarding the example  `hello: "world"` in yaml it's equivalent to `hello: world`, so in less you'll want something complex like multi line string, this should be o.k.

90% of the cases we or need a string or need a number in configurations
this fix helps to handle the case we need a list (and it would also gonna work for mapping)

> 
> https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L832-L841
> 
> and
> https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L872-L875
> 
> or do we accept `foo: [1, False, "hello"]` ?

this should work

> 
> actually, i was about to suggest using `json.loads()` but i just realized that we are not actually expecting json encoded data here. it's kind of freestyle format. but i still think it would be better to stick to a well-defined format / encoding, so we don't need to worry about these things.
> 
> but if we don't really care. i'd suggest at least handle more types of exceptions, see https://docs.python.org/3/library/ast.html#ast.literal_eval

I rather not handle those, let it explode, i.e. if user put in a giant literal or that complex one, that it would hit recursion limit, I don't want to handle it gracefully, and silently put it as a string (which is probably never what he would expect in this case).

Reply:
> I think
> 
> > what if we have something like: `hello: "world"` here? before this change, the `val` is `'"world"'`, after this change the `val` would be `'world'`. and what if we really want to set a value like `'1e10'`. i still cannot tell if this is a valid concern after reading
> 
> regarding the example `hello: "world"` in yaml it's equivalent to `hello: world`, so in less you'll want something complex like multi line string, this should be o.k.

how does yaml come into play? are we consuming `val` later with `yaml.load()` ?

> 
> 90% of the cases we or need a string or need a number in configurations this fix helps to handle the case we need a list (and it would also gonna work for mapping)
> 
> > https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L832-L841
> > 
> > and
> > https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L872-L875
> > 
> > or do we accept `foo: [1, False, "hello"]` ?
> 
> this should work

it works with this patch. but again, do we accept it? or put another way, do we expect it?

> 
> > actually, i was about to suggest using `json.loads()` but i just realized that we are not actually expecting json encoded data here. it's kind of freestyle format. but i still think it would be better to stick to a well-defined format / encoding, so we don't need to worry about these things.
> > but if we don't really care. i'd suggest at least handle more types of exceptions, see https://docs.python.org/3/library/ast.html#ast.literal_eval
> 
> I rather not handle those, let it explode, i.e. if user put in a giant literal or that complex one, that it would hit recursion limit, I don't want to handle it gracefully, and silently put it as a string (which is probably never what he would expect in this case).

if we want to fail the test or whatever using ccm if user feeds it with something malformed, why are we handling `ValueError`? because it is a more typical use case?

Reply:
> > > I think
> > > > what if we have something like: `hello: "world"` here? before this change, the `val` is `'"world"'`, after this change the `val` would be `'world'`. and what if we really want to set a value like `'1e10'`. i still cannot tell if this is a valid concern after reading
> > > 
> > > 
> > > regarding the example `hello: "world"` in yaml it's equivalent to `hello: world`, so in less you'll want something complex like multi line string, this should be o.k.
> > 
> > 
> > how does yaml come into play? are we consuming `val` later with `yaml.load()` ?
> 
> no, if we put it as string in a yaml file, with or without the quotes, it would be parsed the same, the code at scylla end gonna get it without quotes (like any yaml parser would).

i am trying to understand what we are expecting. as there is no canonical document on this. if we want to parse a yaml, then we should go with `yaml.load()` instead of trying to parse it as a python literal. it's both misleading and more importantly, wrong.

> 
> > > 90% of the cases we or need a string or need a number in configurations this fix helps to handle the case we need a list (and it would also gonna work for mapping)
> > > > https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L832-L841
> > > > 
> > > > and
> > > > https://github.com/scylladb/scylla-ccm/blob/69118c7185e5b97a1d923084356f5b4dc4fbec8d/ccmlib/cmds/cluster_cmds.py#L872-L875
> > > > 
> > > > or do we accept `foo: [1, False, "hello"]` ?
> > > 
> > > 
> > > this should work
> > 
> > 
> > it works with this patch. but again, do we accept it? or put another way, do we expect it?
> 
> it's a valid python list, it's a valid json, and would map to valid yaml, I see no reason to block it.
> 
> > > > actually, i was about to suggest using `json.loads()` but i just realized that we are not actually expecting json encoded data here. it's kind of freestyle format. but i still think it would be better to stick to a well-defined format / encoding, so we don't need to worry about these things.
> > > > but if we don't really care. i'd suggest at least handle more types of exceptions, see https://docs.python.org/3/library/ast.html#ast.literal_eval
> > > 
> > > 
> > > I rather not handle those, let it explode, i.e. if user put in a giant literal or that complex one, that it would hit recursion limit, I don't want to handle it gracefully, and silently put it as a string (which is probably never what he would expect in this case).
> > 
> > 
> > if we want to fail the test or whatever using ccm if user feeds it with something malformed, why are we handling `ValueError`? because it is a more typical use case?
> 
> because the the main case, passing some string, which isn't python literal. and not that I'm trying it, this main case of just passing a string is broken

Reply:
> i am trying to understand what we are expecting. as there is no canonical document on this. if we want to parse a yaml, then we should go with yaml.load() instead of trying to parse it as a python literal. it's both misleading and more importantly, wrong.

There is a command line option to update configuration, 
passing a whole yaml structure doesn't yield itself to a command line very easily

Now it's the support configuration option which are string or int, now we need a list.

Reply:
> i am trying to understand what we are expecting. as there is no canonical document on this. if we want to parse a yaml, then we should go with yaml.load() instead of trying to parse it as a python literal. it's both misleading and more importantly, wrong.

There is a command line option to update configuration, 
passing a whole yaml structure doesn't yield itself to a command line very easily

Now it's the support configuration option which are string or int, now we need a list.

Reply:
> There is a command line option to update configuration,

i did read the code implementing  the configuration and referenced it in the previous comment. not sure if that's the one you are referring here?

```
In [1]: import yaml
In [2]: yaml.load('aa')
Out[2]: 'aa'
```

i think "aa" is also a yaml structure, and that's exactly we are passing to the command as the "val" in it.

Reply:
I thought you are talking about parsing whole commandline as yaml...
but yes using `yaml.safe_load` instead of `ast.literal_eval`, works a bit better, since you don't need escaping.

I also added a unit test, to cover the different cases this should be supporting

Reply:
thanks for adding the test. the reason that i prefer `yaml.safe_load` is that we are not loading a python literal. it's more correct, not just more convenient.